📖 Documentation Update to Creating a ROSA HCP cluster directions (#5920)
kubernetes-sigs/cluster-api-provider-aws: docs/book/src/topics/rosa/creating-a-cluster.md

    ```shell
    rosa whoami
    ```

## Permissions

1. Create a new kubernetes secret with the service account credentials to be referenced later by the `ROSAControlPlane`

    ```shell
    kubectl create secret generic rosa-creds-secret \
      --from-literal=ocmClientID='....' \
      --from-literal=ocmClientSecret='eyJhbGciOiJIUzI1NiIsI....' \
      --from-literal=ocmApiUrl='https://api.openshift.com'
    ```

    **Note:** The secret must be created in the same namespace where your ROSA resources will be deployed. Alternatively, to consume the secret without the need to reference it from your `ROSAControlPlane`, name your secret `rosa-creds-secret` and create it in the CAPA manager namespace (usually `capa-system`)

    ```shell
    kubectl -n capa-system create secret generic rosa-creds-secret \
      --from-literal=ocmClientID='....' \
      --from-literal=ocmClientSecret='eyJhbGciOiJIUzI1NiIsI....' \
      --from-literal=ocmApiUrl='https://api.openshift.com'
    ```

## Creating the cluster

1. Prepare the environment:
    ```bash
    export OPENSHIFT_VERSION="4.20.11"  # check available versions with: rosa list versions --hosted-cp
    export AWS_REGION="us-west-2"

>     export AWS_AVAILABILITY_ZONE="us-west-2a"

Reviewer: we could probably provide the `aws sts` commands here:
```
export AWS_ACCOUNT_ID=$(aws sts get-caller-identity --query Account --output text)
export AWS_CREATOR_ARN=$(aws sts get-caller-identity --query Arn --output text)
```

    export AWS_ACCOUNT_ID="<account_id>"
    export AWS_CREATOR_ARN="<user_arn>" # can be retrieved e.g. using `aws sts get-caller-identity`
    ```

1. Create the `ROSARoleConfig` and `ROSANetwork` resources.

    The `ROSARoleConfig` automates the creation of the AWS IAM resources required by ROSA HCP clusters:
    - **Account roles**: Installer, Support, and Worker IAM roles (e.g. `<prefix>-HCP-ROSA-Installer-Role`)
    - **Operator roles**: IAM roles for cluster operators including ingress, image registry, storage, network, kube cloud controller, node pool management, control plane operator, and KMS provider
    - **OIDC provider**: A managed OpenID Connect provider used for operator role authentication

    The `ROSANetwork` automates the creation of the VPC networking infrastructure via an AWS CloudFormation stack, including:

    - A VPC with the specified CIDR block
    - Public and private subnet pairs for each availability zone
    - Associated networking resources (internet gateway, NAT gateways, route tables)

    **Note:** The `prefix` field has a maximum length of 4 characters.

    Save the following to a file named `rosa-role-network.yaml`:

    ```yaml

>     apiVersion: infrastructure.cluster.x-k8s.io/v1beta2

Reviewer: It would be nice that this yaml file and the ones below have the `cat <<EOF ... >` command so that a user can just copy and paste them

    kind: ROSARoleConfig
    metadata:
      name: "role-config"
      namespace: "capa-system"
    spec:
      accountRoleConfig:
        prefix: "rosa"
        version: "4.20.11"
      operatorRoleConfig:
        prefix: "rosa"
      credentialsSecretRef:
        name: rosa-creds-secret
      oidcProviderType: Managed

>     apiVersion: infrastructure.cluster.x-k8s.io/v1beta2

Reviewer: I believe the `---` separator is missing from here

# Creating a ROSA HCP cluster

## Prerequisites

1. Create a management cluster using the [Quick Start Guide.](https://cluster-api-aws.sigs.k8s.io/quick-start) 

2. Install the required tools and set up the prerequisite infrastructure using the [ROSA Setup guide](https://docs.aws.amazon.com/rosa/latest/userguide/set-up.html).
   
Once these steps are complete, you are ready to create a ROSA HCP cluster.

> ## Authentication

Reviewer 2: Please also add another doc refering to enable IAM auth similar to this [doc](https://github.com/stolostron/cluster-api-installer/blob/main/doc/Enable_iam_roles_capa.md)

Author: We should do this separately.

Author: Added document.

The CAPA controller requires service account credentials to provision ROSA HCP clusters.  

> If you already have a service account, you can skip these steps.

Reviewer 2: ```suggestion
**Note:** If you already have a service account, you can skip these steps.
```

Author: Done.

# Creating a ROSA HCP cluster

## Prerequisites

1. Install the required tools and set up the prerequisite infrastructure using the [ROSA Setup guide](https://docs.aws.amazon.com/rosa/latest/userguide/set-up.html).

2. Create a management cluster using the [Quick Start Guide.](https://cluster-api-aws.sigs.k8s.io/quick-start)

## IAM Role Configuration

> Configure the IAM role authentication for the CAPA controller following the directions [here](specify-management-iam-role.md).

Reviewer 2: This step should be as an option not a must to do

```suggestion
Optionally to configure the IAM role authentication for the CAPA controller follow the directions [here](specify-management-iam-role.md).
```

    ```

## Creating the cluster

> 1. Save the following to a file `rosa-role-network.yaml`:

Reviewer 2: Add more context on ROSARoleConfig & ROSANetwork . Like the IAM roles that ROSARoleCOnfig creates and OIDC provider, ...etc

Author: Done.

    ```yaml
    apiVersion: infrastructure.cluster.x-k8s.io/v1beta2
    kind: ROSARoleConfig
    metadata:
      name: "role-config"
    spec:
      accountRoleConfig:
        prefix: "rosa"
        version: "4.19.0"
      operatorRoleConfig:
        prefix: "rosa"
      credentialsSecretRef:
        name: rosa-creds-secret
      oidcProviderType: Managed
    ---
    apiVersion: infrastructure.cluster.x-k8s.io/v1beta2
    kind: ROSANetwork
    metadata:
      name: "rosa-vpc"
    spec:
      region: "us-west-2"
      stackName: "rosa-hcp-net"
      availabilityZones:
      - "us-west-2a"
      - "us-west-2b"
      - "us-west-2c"
      cidrBlock: 10.0.0.0/16
    ```

    Apply the manifest:

    ```shell
    kubectl apply -f rosa-role-network.yaml
    ```

    Verify the `ROSARoleConfig` was successfully created. The status should contain the `accountRolesRef`, `oidcID`, `oidcProviderARN` and `operatorRolesRef`:

    ```shell

>     kubectl get rosaroleconfig role-config -o yaml

Reviewer 2: Add example of the expected ROSARoleConfig->Status

Author: Done.

    ```

    Verify the `ROSANetwork` was successfully created. The status should contain the created subnets:

    ```shell

>     kubectl get rosanetwork rosa-vpc -o yaml

Reviewer 2: Add example of the expected ROSANetwork->Status

Author: Done.

    - Associated networking resources (internet gateway, NAT gateways, route tables)

    Save the following to a file `rosa-role-network.yaml`:

    ```yaml
    apiVersion: infrastructure.cluster.x-k8s.io/v1beta2
    kind: ROSARoleConfig
    metadata:
      name: "role-config"
    spec:
      accountRoleConfig:
        prefix: "rosa"
        version: "4.19.0"
      operatorRoleConfig:
        prefix: "rosa"
      credentialsSecretRef:
        name: rosa-creds-secret
      oidcProviderType: Managed
    ---
    apiVersion: infrastructure.cluster.x-k8s.io/v1beta2
    kind: ROSANetwork
    metadata:
      name: "rosa-vpc"
    spec:
      region: "us-west-2"
      stackName: "rosa-hcp-net"
      availabilityZones:
      - "us-west-2a"
      - "us-west-2b"
      - "us-west-2c"
      cidrBlock: 10.0.0.0/16
    ```

    Apply the manifest:

    ```shell
    kubectl apply -f rosa-role-network.yaml
    ```

    Verify the `ROSARoleConfig` was successfully created. The status should contain the `accountRolesRef`, `oidcID`, `oidcProviderARN` and `operatorRolesRef`:

    ```shell
    kubectl get rosaroleconfig role-config -o yaml
    ```

    Example expected status:

    ```yaml

>     status:

Reviewer 2: ```suggestion
    apiVersion: infrastructure.cluster.x-k8s.io/v1beta2
    kind: ROSARoleConfig
    metadata:
      name: "role-config"
    spec:
      ...
    status:
```

Author: Added.

      oidcProviderARN: arn:aws:iam::123456789012:oidc-provider/oidc.os1.devshift.org/anyoidcanyoidctuq4b
      operatorRolesRef:
        controlPlaneOperatorARN: arn:aws:iam::123456789012:role/rosa-kube-system-control-plane-operator
        imageRegistryARN: arn:aws:iam::123456789012:role/rosa-openshift-image-registry-installer-cloud-credentials
        ingressARN: arn:aws:iam::123456789012:role/rosa-openshift-ingress-operator-cloud-credentials
        kmsProviderARN: arn:aws:iam::123456789012:role/rosa-kube-system-kms-provider
        kubeCloudControllerARN: arn:aws:iam::123456789012:role/rosa-kube-system-kube-controller-manager
        networkARN: arn:aws:iam::123456789012:role/rosa-openshift-cloud-network-config-controller-cloud-credentials
        nodePoolManagementARN: arn:aws:iam::123456789012:role/rosa-kube-system-capa-controller-manager
        storageARN: arn:aws:iam::123456789012:role/rosa-openshift-cluster-csi-drivers-ebs-cloud-credentials
    ```

    Verify the `ROSANetwork` was successfully created. The status should contain the created subnets:

    ```shell
    kubectl get rosanetwork rosa-vpc -o yaml
    ```

    Example expected status:

    ```yaml

>     status:

Reviewer 2: ```suggestion
   apiVersion: infrastructure.cluster.x-k8s.io/v1beta2
    kind: ROSANetwork
    metadata:
      name: "rosa-vpc"
    spec:
       ...
    status:
```

Author: Added.

        publicSubnet: subnet-0bb2220505b16f606
      - availabilityZone: us-west-2c
        privateSubnet: subnet-002e071b9624727f3
        publicSubnet: subnet-049fa2a528d896356
    ```

1. Save the following to a file `rosa-cluster.yaml`:

    ```yaml
    apiVersion: cluster.x-k8s.io/v1beta1
    kind: Cluster
    metadata:
      name: "rosa-hcp-1"
    spec:
      clusterNetwork:
        pods:
          cidrBlocks: ["192.168.0.0/16"]
      infrastructureRef:
        apiVersion: infrastructure.cluster.x-k8s.io/v1beta2
        kind: ROSACluster
        name: "rosa-hcp-1"
      controlPlaneRef:
        apiVersion: controlplane.cluster.x-k8s.io/v1beta2
        kind: ROSAControlPlane
        name: "rosa-hcp-1-control-plane"
    ---
    apiVersion: infrastructure.cluster.x-k8s.io/v1beta2
    kind: ROSACluster
    metadata:
      name: "rosa-hcp-1"
    spec: {}
    ---
    apiVersion: controlplane.cluster.x-k8s.io/v1beta2
    kind: ROSAControlPlane
    metadata:
      name: "rosa-hcp-1-control-plane"
    spec:
      credentialsSecretRef:
        name: rosa-creds-secret
      rosaClusterName: rosa-hcp-1
      domainPrefix: rosa-hcp
      rosaRoleConfigRef:

>         name: role-config

Reviewer 2: ```suggestion
        name: role-config  # reference to the ROSARoleConfig created above
```

Author: Done.

      version: "4.19.0"
      region: "us-west-2"
      rosaNetworkRef:

>         name: "rosa-vpc"

Reviewer 2: ```suggestion
        name: "rosa-vpc" # reference to the ROSANetwork created above
```

Author: Done.

    spec:
      allowedNamespaces: {}  # matches all namespaces
    ```
    Apply the manifest:

    ```shell
    kubectl apply -f aws-cluster-controller-identity.yaml
    ```

1. Save the following to a file named `rosa-cluster.yaml`:

    ```yaml
    apiVersion: cluster.x-k8s.io/v1beta2
    kind: Cluster
    metadata:
      name: "rosa-hcp-1"
    spec:
      clusterNetwork:
        pods:
          cidrBlocks: ["192.168.0.0/16"]
      infrastructureRef:
        apiGroup: infrastructure.cluster.x-k8s.io
        kind: ROSACluster
        name: "rosa-hcp-1"
      controlPlaneRef:
        apiGroup: controlplane.cluster.x-k8s.io
        kind: ROSAControlPlane

>         name: "rosa-hcp-1-control-plane"

Reviewer: I believe the --- separator is missing from here

    apiVersion: infrastructure.cluster.x-k8s.io/v1beta2
    kind: ROSACluster
    metadata:
      name: "rosa-hcp-1"

>     spec: {}

Reviewer: I believe the --- separator is missing from here

    ```shell
    kubectl get ROSAControlPlane rosa-hcp-1-control-plane

    NAME                       CLUSTER      READY
    rosa-hcp-1-control-plane   rosa-hcp-1   true
    ```

    The ROSA HCP cluster can take around 40 minutes to be fully provisioned.

1. After provisioning has completed, verify the `ROSAMachinePool` resources were successfully created:

    ```shell
    kubectl get ROSAMachinePool

    NAME        READY   REPLICAS
    workers-0   true    1
    workers-1   true    1
    workers-2   true    1
    ```

    **Note:** The number of default `ROSAMachinePool` resources corresponds to the number of availability zones configured.

1. To add an additional `ROSAMachinePool`, save the following to a file `rosa-machinepool-extra.yaml`:

    ```yaml
    apiVersion: cluster.x-k8s.io/v1beta2
    kind: MachinePool
    metadata:
      name: "rosa-hcp-1-workers-extra"
    spec:
      clusterName: "rosa-hcp-1"
      replicas: 2
      template:
        spec:
          clusterName: "rosa-hcp-1"
          bootstrap:
            dataSecretName: ""
          infrastructureRef:
            apiGroup: infrastructure.cluster.x-k8s.io
            kind: ROSAMachinePool

>             name: "workers-extra"

Reviewer: missing `---` separator here

      allowedNamespaces: {}  # matches all namespaces
    ```

    ```shell
    kubectl apply -f <filename>.yaml
    ```

    see [Multi-tenancy](../multitenancy.md) for more details

1. Check the `ROSAControlPlane` status:

    ```shell
    kubectl get ROSAControlPlane rosa-hcp-1-control-plane

    NAME                       CLUSTER      READY
    rosa-hcp-1-control-plane   rosa-hcp-1   true
    ```

    The ROSA HCP cluster can take around 40 minutes to be fully provisioned.

1. After provisioning has completed, verify the `ROSAMachinePool` resources were successfully created:

    ```shell
    kubectl get ROSAMachinePool

    NAME        READY   REPLICAS
    workers-0   true    1
    workers-1   true    1
    workers-2   true    1
    ```

    **Note:** The number of default `ROSAMachinePool` resources corresponds to the number of availability zones configured.

> 

Reviewer 2: May add extra step here for adding ROSAMachinePool ex; workers-extra

Reviewer 2: Also add section for deleting a ROSA-HCP cluster

Author: Added.

1. To add an additional `ROSAMachinePool`, save the following to a file `rosa-machinepool-extra.yaml`:

    ```yaml
    apiVersion: cluster.x-k8s.io/v1beta1
    kind: MachinePool
    metadata:
      name: "rosa-hcp-1-workers-extra"
    spec:
      clusterName: "rosa-hcp-1"
      replicas: 2
      template:
        spec:
          clusterName: "rosa-hcp-1"
          bootstrap:
            dataSecretName: ""
          infrastructureRef:
            apiVersion: infrastructure.cluster.x-k8s.io/v1beta2
            kind: ROSAMachinePool
            name: "workers-extra"
    ---
    apiVersion: infrastructure.cluster.x-k8s.io/v1beta2
    kind: ROSAMachinePool
    metadata:
      name: "workers-extra"
    spec:
      nodePoolName: "workers-extra"
      version: "4.19.0"
      instanceType: "m5.xlarge"
      autoRepair: true
    ```

    ```shell
    kubectl apply -f rosa-machinepool-extra.yaml
    ```

## Deleting a ROSA HCP cluster

To delete a ROSA HCP cluster, delete the `Cluster` resource. This will cascade and delete all associated resources including the `ROSAControlPlane`, `ROSACluster`, and `ROSAMachinePool` resources:

```shell

> kubectl delete cluster rosa-hcp-1

Reviewer 2: this is not correct , it should delete the rosacontrolplane as well

Author: Modified.
